contrib/natsio/src/main/java/com/google/cloud/dataflow/contrib/natsio/NatsIO.java

```diff
@@ -0,0 +1,395 @@
+		 * @param props contains NATS related properties. "servers" property is mandatory to access a remote NATS server.
+		 * @return a bounded NatsSource instance
+		 */
+		public static PTransform<PInput, PCollection<KV<String,String>>> withMaxNumRecords(String subject, long numRecords, Properties props) {
+			return from(subject, props).withMaxNumRecords(numRecords);
+		}		
+
+		/**
+		 * Create a NatsSource instance which is bounded to total reading time to receive updates.
+		 * @param subject to subscribe and received updates
+		 * @param readTime bounds total reading time to receive updates from NATS server.
+		 * @param props contains NATS related properties. "servers" property is mandatory to access a remote NATS server.
+		 * @return a bounded NatsSource instance
+		 */
+		public static PTransform<PInput, PCollection<KV<String,String>>> withMaxReadtime(String subject, long readTime, Properties props) {
+			return from(subject, props).withMaxReadTime(Duration.standardSeconds(readTime));
+		}
+		
+		public static class NatsSource extends UnboundedSource<KV<String,String>, UnboundedSource.CheckpointMark> {
+			private static final long serialVersionUID = 1L;
+			
+			private Properties props = null;
+			private Connection conn = null;
+			private String subject;
+			
+			public NatsSource(String subject, Properties props) {
+				if (subject == null) {
+					throw new NullPointerException("subject");
+				}
+				if (props == null) {
+					throw new NullPointerException("props");
+				}
+				if (props.getProperty("servers") == null) {
+					throw new NullPointerException("servers");
+				}
+				this.subject = subject;
+				this.props = props;
+			}
+			
+			public Properties getProperties() {
+				return props;
+			}
+			
+			public Connection getConnection() {
+				return conn;
+			}
+			
+			public String getSubject() {
+				return subject;
+			}
+			
+			@Override
+			public NatsReader createReader(PipelineOptions options, UnboundedSource.CheckpointMark mark) {
+				return new NatsReader(subject);
+			}
+
+			@Override
+			public List<? extends UnboundedSource<KV<String,String>, UnboundedSource.CheckpointMark>> generateInitialSplits(
+					int numSplits, PipelineOptions options) throws Exception {
+				List<NatsSource> results = new ArrayList<NatsSource>();
+				// Create a NatsSource instance for each split.
+				for(int i = 0; i < numSplits; i++) {
+					results.add(new NatsSource(subject, props));
+				}
+				return results;
+			}
+
+			@Override
+			public Coder<UnboundedSource.CheckpointMark> getCheckpointMarkCoder() {
+				return null;
+			}
+
+			@Override
+			public Coder<KV<String,String>> getDefaultOutputCoder() {
+				return KvCoder.of(NatsIO.DEFAULT_CODER, NatsIO.DEFAULT_CODER);
+			}
+
+			@Override
+			public void validate() {
+				Preconditions.checkNotNull(props.getProperty("servers"), "need to set the 'servers' property");
+				Preconditions.checkNotNull(subject, "need to set the 'subject'");
```

Comment:
is there a way these can be `null` or missing given how the source is constructed?